---
title: "Motif Scanning using FIMO"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Motif Scanning using FIMO}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(dremeR)
```

## Inputs
 - sequences (list not supported since no reason to split, instead use plyranges joins after fimo to add metadata)
  - copy info from Chip-seq example
  - explain about why `get_sequence()` is good to use, becaues of `parse_genomic_coord`
 - motifs: path or universalmotif
 
## Note about settings 
  - I use nondefault fimo settings to make it work better with R interface and run faster.
  - text = TRUE
    - if running on huge sequence, may overflow memory. Link to article about running FIMO on whole genome.

## Saving data from FIMO Web Server
![saving FIMO tsv results](save_fimo.png)

# Citation

dremeR is a wrapper for a select few tools from the MEME Suite, which were
developed by another group. In addition to citing dremeR, please cite the MEME
Suite tools corresponding to the tools you use.

If you use `runFimo()` in your analysis, please cite:

Charles E. Grant, Timothy L. Bailey, and William Stafford Noble, "FIMO: Scanning for occurrences of a given motif", Bioinformatics, 27(7):1017-1018, 2011. [full text](http://bioinformatics.oxfordjournals.org/content/early/2011/02/16/bioinformatics.btr064.full)
