---
title: "ChIP-seq Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Integrative Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
NOT_CRAN <- identical(tolower(Sys.getenv("NOT_CRAN")), "true")

eval_vignette <- NOT_CRAN & memes::meme_is_installed()

knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  purl = eval_vignette,
  eval = eval_vignette
)
```

```{r setup}
library(memes)
library(magrittr)
suppressPackageStartupMessages(library(GenomicRanges))
```

```{r}
check_meme_install()
```

## Experimental Design

### Prepare peaks for analysis
```{r, message=F}
data("example_chip_summits", package = "memes")
```

```{r}
# These data use the dm3 reference genome
dm.genome <- BSgenome.Dmelanogaster.UCSC.dm3::BSgenome.Dmelanogaster.UCSC.dm3

# Get sequences in a 100bp window around the peak summit
summit_flank <- example_chip_summits %>% 
  plyranges::anchor_center() %>% 
  plyranges::mutate(width = 100)
```

## Determinants of ectopic and orphan binding

### Pre-filtering database for expressed transcription factors
```{r}
meme_db_path <- system.file("extdata/flyFactorSurvey_cleaned.meme", package = "memes")
```

```{r}
library(universalmotif)
meme_db <- read_meme(meme_db_path) %>% 
  as_universalmotif_dataframe()
```

```{r}
data("example_rnaseq", package = "memes")
```

```{r, fig.height=5, fig.width=5}
library(ggplot2)

example_rnaseq %>% 
  dplyr::group_by(symbol) %>% 
  dplyr::filter(fpkm == max(fpkm)) %>% 
  ggplot(aes(fpkm)) +
    geom_histogram(binwidth = 1) +
    coord_cartesian(xlim = c(0, 25)) +
    theme_bw() +
    labs(title = "Largest FPKM value in RNAseq dataset for each gene",
         y = "Number of genes",
         x = "Largest FPKM Value")
```

```{r}
expressed_genes <- example_rnaseq %>% 
  # For each gene, keep only those with max(FPKM) greater or equal to 5.
  dplyr::group_by(symbol) %>% 
  dplyr::filter(max(fpkm) >= 5)
```

```{r}
meme_db_expressed <- meme_db %>% 
  # the altname slot of meme_db contains the gene symbol
  # (this is database-specific)
  dplyr::filter(altname %in% expressed_genes$symbol)
```

```{r}
#Number of motifs pre-filtering: 
nrow(meme_db)
#Number of motifs post-filter: 
nrow(meme_db_expressed)
```

We've decreased our multiple-testing penalty by ~55%.

### Examination of binding categories with AME
```{r}
by_binding <- summit_flank %>% 
  split(mcols(.)$peak_binding_description) %>% 
  get_sequence(dm.genome)
```

```{r}
options(meme_db = as_universalmotif(meme_db_expressed))
```

```{r}
ame_by_binding <- by_binding %>% 
  runAme
```

### Visualizing AME results 

The `ame_plot_heatmap()` function provides a quick way to visualize AME results.
It is built on top of `ggplot2`, so all `ggplot2` functions can be used to
further modify the plot.

By default, it uses the -log10(adjusted p-value) as the heat values. See the
documentation for additional notes on customization.
```{r, fig.height=3.5, fig.width=7}
ame_by_binding$ectopic %>% 
  dplyr::filter(rank %in% 1:10) %>% 
  ame_plot_heatmap(group_name = "Ectopic Sites")  +
    ggtitle("Top 10 AME Hits in Ectopic Sites")
```

To plot results from multiple runs together, they must first be joined into 1
data frame. The `ame_by_binding` object is a list whose names correspond to the
E93 binding category. The list can be combined into a data.frame using
`dplyr::bind_rows`. Setting `.id = "binding_type` creates a new column
`binding_type` that contains the names from the `ame_by_binding` list. In this
way, the `ame_res` data.frame contains all AME results for each run, which can
be distinguished by the `binding_type` column.
```{r}
ame_res <- ame_by_binding %>% 
  dplyr::bind_rows(.id = "binding_type")
  
```

It is possible to aggregate results from multiple runs into a heatmap by setting
the `group` parameter in `ame_plot_heatmap()`.

```{r, fig.height=5,fig.width=15}
ame_res %>% 
  ame_plot_heatmap(group = binding_type)
```

### Reducing redundant motif hits
There is another key consideration for the above visualization and that is that
in the FlyFactorSurvey database we used, different TFs can have multiple motif
entries in the database which are all detected separately by AME. Here, when
returning the top 5 hits from each group, you can see, for example, that a motif
matching "Aef1" is reported 2 times within the top 5 hits of the ectopic sites
(A similar result can also be seen for motifs matching "klu").
In this situation, it makes sense to summarize the data at the TF level, instead
of the motif level. **Note:** There may be exceptions to this if for example, a
TF has multiple DNA binding sequences it can recognize, in which case having
multiple hits may reflect a biological property of your sequences. You will have
to handle this on a case-by-case basis for interesting hits and different motif
databases. Here, we can see that at least for Aef1 and klu, the consensus
sequences are very similar.
```{r}
ame_res %>% 
  dplyr::filter(binding_type == "ectopic", rank %in% 1:5) %>% 
  head(5) %>% 
  dplyr::select(binding_type, rank, motif_id, motif_alt_id, consensus)
```

How to solve this problem will vary with different motif databases (For details
on how to pre-process a motif database, see `vignette("tidy_motifs")`). In the
tidy version of the FlyFactorSurvey database, the `altname`s of the motifs are
set to the transcription factor gene symbol. This information is included in the
AME results as the `motif_alt_id` column.

```{r, fig.height=3, fig.width=12}
ame_res %>% 
  dplyr::group_by(binding_type, motif_alt_id) %>% 
  dplyr::filter(adj.pvalue == min(adj.pvalue)) %>% 
  ame_plot_heatmap(group = binding_type, id = motif_alt_id) +
    labs(y = "Binding Category",
         x = "Transcription Factor Motif")
```


### AME Heatmap Visualization
If you have read `vignette("core_ame")`, the "normalized rank" heatmap was
introduced as an alternative visualization method useful when AME produces very
large numbers of hits, or when p-values are on very different scales between groups. 
The `ame_compare_heatmap_methods()` function can be used to visualize why.

Below is a comparison of the distribution of values when using
`-log10(adj.pvalue)` (A) vs normalized ranks (B). Because there are relatively
few hits in the results (~30), and the number of hits between groups varies more
than the `-log10(p-value)` distributions, the "normalize" method will produce a misleading
heatmap relative to the `-log10(p-value)` map.

```{r, fig.height=3, fig.width=7.5}
ame_res %>% 
  dplyr::group_by(binding_type, motif_alt_id) %>% 
  dplyr::filter(adj.pvalue == min(adj.pvalue)) %>% 
  ame_compare_heatmap_methods(group = binding_type)
```

```{r, fig.height=3, fig.width=12}
best_ame_hits <- ame_res %>% 
  dplyr::group_by(binding_type, motif_alt_id) %>% 
  dplyr::filter(adj.pvalue == min(adj.pvalue))
  
pval_heatmap <- best_ame_hits %>% 
  ame_plot_heatmap(group = binding_type, id = motif_alt_id) +
    labs(x = NULL,
         title = "-log10(adj.pvalue)")
  

norm_heatmap <- best_ame_hits %>% 
  ame_plot_heatmap(group = binding_type, id = motif_alt_id, value = "normalize") +
    labs(x = NULL,
         title = "normalize")

pval_scaled_heatmap <- best_ame_hits %>% 
  ame_plot_heatmap(group = binding_type, id = motif_alt_id, scale_max = 10) +
    labs(x = NULL,
         title = "-log10(adj.pvalue) (scale capped at 10)")
```

```{r, fig.height=9, fig.width=12}
cowplot::plot_grid(pval_heatmap,
                   norm_heatmap,
                   pval_scaled_heatmap,
                   ncol = 1,
                   labels = "AUTO")
```

## Denovo motif similarity by binding
```{r dreme_by_binding, eval=F}
# TODO: decide whether to keep this chunk or not, depends on build-time of vignettes
dreme_by_binding <- by_binding %>% 
  #TODO: work out this chunk
  #runDreme("shuffle", nmotifs = 5) %>% 
  runDreme("shuffle", nmotifs = 5) %>% 
  dplyr::bind_rows(.id = "binding_type")
```

```{r}
# The above code chunk takes a long time to run.
# memes is packaged with the results of this run in the "example_dreme_by_binding" dataset
# which can be loaded as follows:
data("example_dreme_by_binding", package = "memes")
dreme_by_binding <- example_dreme_by_binding %>% 
  dplyr::bind_rows(.id = "binding_type")
```

Rename the motifs to indicate the binding category they were discovered in.
```{r}
dreme_by_binding_renamed <- dreme_by_binding %>% 
  dplyr::mutate(name = paste(binding_type, seq, sep = "_")) %>% 
  # update_motifs updates the information in the special `motif` column
  update_motifs()
```

```{r, eval=F}
stop("RETURN HERE. pheatmap causes a: readRDS(pfile) : error reading the file") 
# d <- dreme_by_binding_renamed %>% 
#   {universalmotif::compare_motifs(.$motif, method = "PCC")} 

#>   pheatmap::pheatmap(d, color = cols, 
#+                      breaks = seq(0, 1, by = 1/255), 
#+                      annotation_col = anno.df,
#+                      show_colnames = FALSE) 
#error reading the fileError in readRDS(pfile) : error reading from connection
#In addition: Warning messages:
#1: In readRDS(pfile) : error reading the file
#2: In readRDS(pfile) : error reading the file
#3: In readRDS(pfile) : error reading the file
#4: In readRDS(pfile) : error reading the file
```

```{r, fig.height=5, fig.width=8}
cols <- colorRampPalette(c("white", "dodgerblue4"))(255)

anno.df <- dreme_by_binding_renamed %>% 
  dplyr::select(name, binding_type) %>% 
  tibble::column_to_rownames("name")

dreme_by_binding_renamed %>% 
  {universalmotif::compare_motifs(.$motif, method = "PCC")} %>% 
  pheatmap::pheatmap(color = cols, 
                     breaks = seq(0, 1, by = 1/255), 
                     annotation_col = anno.df, 
                     cutree_rows = 6,
                     cutree_cols = 6,
                     show_colnames = FALSE) 
```

### Test Denovo motif enrichment using AME

The above analysis suggests the motif content of the different binding
categories are highly similar. To extend these analyses, we can use AME to test
for motif enrichment of the denovo discovered motifs within each binding category.

`runAme()` allows using a `runDreme()` results object as the `database` input by
passing it within a `list()`. Naming the `list()` entry produces an informative
`motif_db` name in the results data.frame.

```{r ame_by_binding}
ame_denovo_by_binding <- by_binding %>% 
  runAme(database = list("denovo_binding_motifs" = dreme_by_binding_renamed)) %>% 
  dplyr::bind_rows(.id = "binding_type") 
```

Plotting the heatmap of results reveals that indeed a majority of the Denovo
motifs discovered within a single category are detected in all 3 categories,
supporting the conclusion that orphan, ectopic, and entopic sites are highly
similar in sequence content.
```{r, fig.height=4, fig.width=10}
ame_denovo_by_binding %>% 
  ame_plot_heatmap(group = binding_type)
```
```{r, fig.height=3, fig.width=7.5}
ame_denovo_by_binding %>% 
  ame_compare_heatmap_methods(group = binding_type)
```
```{r, fig.height=4, fig.width=10}
ame_denovo_by_binding %>% 
  ame_plot_heatmap(group = binding_type, value = "normalize")
```

However, there are 4 interesting motifs which distinguish orphan and ectopic
sites from entopic sites. To help identify which TFs these motifs might belong
to, we can use TomTom to match them to known motifs.

First, we isolate the motifs by ID which are unique to 1 group.
```{r}
ame_denovo_binding_unique <- ame_denovo_by_binding %>% 
  # Counts the number of groups a motif is found within
  dplyr::group_by(motif_id) %>% 
  dplyr::add_count() %>% 
  # Selects only rows from motifs found in 1 group
  dplyr::filter(n == 1)
```

Next, we use the motif id's from the AME results to filter the dreme results, and run TomTom on that subset.
```{r}
dreme_by_binding_unique <- dreme_by_binding_renamed %>% 
  dplyr::filter(name %in% ame_denovo_binding_unique$motif_id) %>% 
  runTomTom(dist = "ed")
```

Finally, we visualize the TomTom results to identify candidate TFs driving the presence of ectopic and orphan sites.
```{r, fig.height=4, fig.width=12}
dreme_by_binding_unique %>% 
  view_tomtom_hits(3) %>% 
  cowplot::plot_grid(plotlist = ., nrow = 1, labels = "AUTO")
```

## Motifs in opening vs closing sites
```{r}
# split by response to E93 binding
by_sens <- summit_flank %>% 
  split(mcols(.)$e93_sensitive_behavior) %>% 
  get_sequence(dm.genome)
```

```{r dreme_by_sens_vs_static}
dreme_by_sens_vs_static <- runDreme(by_sens, "Static")
```

```{r}
dreme_results <- dreme_by_sens_vs_static %>% 
  dplyr::bind_rows(.id = "e93_response")
```

```{r}
all_res <- dreme_results %>% 
  runTomTom(dist = "ed")
```

```{r, fig.height=8, fig.width=8}
all_res %>% 
  view_tomtom_hits(3) %>% 
  cowplot::plot_grid(plotlist = ., labels = "AUTO")
```
```{r}
# use force_best_match to use update the best match info
all_res %<>% 
  force_best_match(c("m01_CSAAAAM" = "Eip93F_SANGER_10",
                     "m03_AKGG" = "pho_SANGER_10"))
```


```{r, fig.height=4.5, fig.width=8}
all_res %>% 
  view_tomtom_hits(1) %>% 
  cowplot::plot_grid(plotlist = ., labels = "AUTO")
```

```{r, fig.height=3, fig.width=8}
all_res %>% 
  dplyr::mutate(label = paste0(e93_response, " in response to E93")) %>% 
  split(.$label) %>% 
  purrr::imap(~{
    top_hits <- view_tomtom_hits(.x, 1) %>% 
      cowplot::plot_grid(plotlist = ., nrow = 1, labels = "AUTO")
    
    title <- cowplot::ggdraw() +
        cowplot::draw_text(.y)
    
    cowplot::plot_grid(plotlist = list(title, top_hits), 
                       ncol = 1,
                       rel_heights = c(0.1, 1)
    )
  })
```


### FIMO (by response)
Because we discover a *de-novo* motif from DREME that matches to E93, it can be useful to scan peaks for motif occurrances to further examine properties of different peak categories.

Although we could also import the motifs from our local .meme database file, it is also possible to use motifs pulled from a [`MotifDb`](https://bioconductor.org/packages/release/bioc/html/MotifDb.html) query as follows. 
```{r}
e93_flyfactor <- MotifDb::MotifDb %>% 
  # Query the database for the E93 motif using it's gene name
  MotifDb::query("Eip93F") %>% 
  # Convert from motifdb format to universalmotif format
  universalmotif::convert_motifs() %>% 
  # The result is a list, to simplify the object, return it as a single universalmotif
  .[[1]]

# Rename the motif from it's flybase gene number to a more user-friendly name
e93_flyfactor["name"] <- "E93_FlyFactor"

e93_flyfactor
```

Just like `runDreme()` and `runAme()`, `runFimo()` takes a
`Biostrings::XStringSet` as input. Using `get_sequence()` to generate these for
DNA sequences is preferred because it creates sequence inputs that can be parsed
into genomic coordinates by Fimo.
```{r}
fimo_res <- summit_flank %>% 
  get_sequence(dm.genome) %>% 
  runFimo(motifs = e93_flyfactor, thresh = 1e-3)
```

The results of `runFimo()` are returned as a `GRanges` object containing the
positions of each motif discovered in the input sequences. The best way to
integrate these data with our input peaks is to use the `plyranges` suite of
tools for performing overlaps and joins between `GRanges` objects.
```{r}
fimo_res
```

### Counting the number of motifs per peak

`plyranges` extends `dplyr`-like syntax to range objects. Here we add a count
for each motif per peak in the `n_motifs` column. We also add a column
`has_motif` which will be a binary indicator of whether a peak contains any
motifs.
```{r}
summit_flank %<>% 
  # Note, if running FIMO with multiple motifs, this solution will not work
  # as it will count all motifs within the fimo-results without splitting by motif_id
  plyranges::mutate(n_motifs = plyranges::count_overlaps(., fimo_res), 
                    has_motif = n_motifs > 0)
```

First, to we want to determine whether E93 sensitive sites are more likely to
have E93 motifs in certain response types. Here we can see that sensitive
decreasing sites are more likely to have E93 motifs than sensitive increasing or
insensitive static sites.
```{r, fig.height=4, fig.width=5}
summit_flank %>% 
  data.frame %>% 
  dplyr::mutate(has_match = dplyr::if_else(has_motif, "Match", "No Match")) %>% 
  ggplot(aes(e93_sensitive_behavior)) +
    geom_bar(aes(fill = forcats::fct_rev(has_match)), position = "fill") +
    scale_fill_manual(values = c("Match" = "firebrick",
                                 "No Match" = "Black")) +
    labs(fill = "E93 Motif Match",
         y = "Fraction of Sites",
         x = "Response to E93 binding")
```


To investigate whether E93-sensitive sites that open/close/don't change have
different numbers of E93 motifs, we plot the fraction of sites with each number
of motifs. Here we can see that in addition to being more likely to contain an
E93 motif, sensitive decreasing sites are more likely to to contain 2 or more
matches, where 10% contain at least 2 motifs.
```{r, fig.height=5, fig.width=7}
summit_flank %>% 
  # currently, group operations are faster as data.frames, so we convert to data.frame
  data.frame %>% 
  dplyr::group_by(e93_sensitive_behavior, n_motifs) %>% 
  dplyr::count() %>% 
  dplyr::group_by(e93_sensitive_behavior) %>% 
  dplyr::mutate(frac = n/sum(n)) %>% 
  ggplot(aes(n_motifs, frac)) +
    geom_line(aes(color = e93_sensitive_behavior), size = 1) +
    labs(y = "Fraction of Sites",
         x = "Number of E93 Motifs",
         color = "Response to E93 Binding") +
    theme_bw()
```

Finally, we want to assess whether the quality of E93 motifs is different between sensitivity categories. To examine this, we need to determine which motifs are found in which peaks. We use the `plyranges::join_overlap_intersect` to return motif entries appended with peak-level metadata, like the peak id each motif is found within.
```{r}
# return position of each motif match w/ peak metadata
intersect <- fimo_res %>% 
  plyranges::join_overlap_intersect(summit_flank)
```

We use the FIMO `score` as a proxy for quality, where higher scores are better
matches to the motif. Here we examine only the best match (highest score) motif
within each peak. 
```{r}
best_motifs <- intersect %>% 
  plyranges::group_by(id) %>% 
  # Keep only motifs w/ the highest score within a peak
  plyranges::filter(score == max(score)) %>% 
  plyranges::ungroup()
```

```{r, fig.height=4, fig.width=3}
best_motifs %>% 
  data.frame %>% 
  ggplot(aes(e93_sensitive_behavior, score)) +
    geom_boxplot(aes(fill = e93_sensitive_behavior), notch = TRUE, width = 0.5) +
    guides(fill = "none") +
    labs(x = "Response to E93",
         y = "FIMO Score") +
    theme_bw()
```

The [Anderson-Darling test](https://en.wikipedia.org/wiki/Anderson%E2%80%93Darling_test) can be used to test differences in the distribution of FIMO scores across multiple groups. For a pairwise comparison, a [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) is commonly used instead.
```{r}
best_motifs %>% 
  data.frame %>% 
  dplyr::mutate(behavior = factor(e93_sensitive_behavior)) %>% 
  PMCMRplus::adAllPairsTest(score ~ behavior, ., p.adjust.method = "fdr")
```

We can visualize the differences in E93 motif quality between groups by
reconstructing PCM's from the matched motif sequences. Here, it is clear that
although each category contains an E93 motif match, the sequence content of
those motifs differs between categories, providing a nice visual representation
of the statistical analysis above.
```{r, fig.height=6, fig.width=5}
best_motifs %>% 
  # Get sequences of each matched motif
  add_sequence(dm.genome) %>% 
  # Split by E93 response category
  split(mcols(.)$e93_sensitive_behavior) %>% 
  # Convert from GRangesList to list() to use `purrr::imap`
  as.list() %>% 
  # imap passes the list entry as .x and the name of that object to .y
  purrr::imap(~{
    # run create_motif() on each set of sequences
    create_motif(.x$sequence, 
                 # Append the response category to the motif name
                 name = paste0("E93_", .y))
    }) %>% 
  # Prepend E93 Fly Factor motif to beginning of motif list
  c(list("E93_FlyFactor" = e93_flyfactor), .) %>% 
  # Plot each motif
  view_motifs()
```


### Centrality of E93 motif
Next we want to visualize whether the E93 motif has altered positioning across
the response categories. To do this we add the metadata for each nearest motif
to our peak summits. Setting `distance = TRUE` in `plyranges::join_nearest` adds
a `distance` column indicating the distance to the nearest joined region.
```{r}
summit_nearest_e93_motif <- summit_flank %>% 
  plyranges::anchor_center() %>% 
  plyranges::mutate(width = 1) %>% 
  plyranges::join_nearest(fimo_res, distance = TRUE)
```

```{r}
summit_nearest_e93_motif %>% 
  data.frame %>% 
  dplyr::filter(has_motif == TRUE) %>% 
  ggplot(aes(distance)) +
    stat_ecdf(aes(color = e93_sensitive_behavior), size = 1, pad = FALSE) +
    labs(x = "Distance to nearest E93 Motif",
         y = "Fraction of Sites",
         color = "E93 Response") +
    theme_linedraw() 
```
```{r}
summit_nearest_e93_motif %>% 
  data.frame %>% 
  dplyr::filter(has_motif == TRUE) %>% 
  dplyr::mutate(behavior = factor(e93_sensitive_behavior)) %>% 
  # Anderson-darling test for multiple distributions
  PMCMRplus::adAllPairsTest(distance ~ behavior, ., p.adjust.method = "fdr") 

```
