<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ChIP-seq Analysis • memes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ChIP-seq Analysis">
<meta property="og:description" content="memes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/memes/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ChIP-seq Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/memes/blob/master/vignettes/integrative_analysis.Rmd"><code>vignettes/integrative_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>integrative_analysis.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">memes</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">GenomicRanges</span>))</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="../reference/check_meme_install.html">check_meme_install</a></span>()
<span class="co">#&gt; checking main install</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin</span>
<span class="co">#&gt; checking util installs</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin/dreme</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin/ame</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin/fimo</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin/tomtom</span>
<span class="co">#&gt; ✔ /nas/longleaf/home/snystrom/meme/bin/meme</span></pre></body></html></div>
<div id="experimental-design" class="section level2">
<h2 class="hasAnchor">
<a href="#experimental-design" class="anchor"></a>Experimental Design</h2>
<div id="prepare-peaks-for-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-peaks-for-analysis" class="anchor"></a>Prepare peaks for analysis</h3>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"example_chip_summits"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"memes"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># These data use the dm3 reference genome</span>
<span class="no">dm.genome</span> <span class="kw">&lt;-</span> <span class="kw pkg">BSgenome.Dmelanogaster.UCSC.dm3</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Dmelanogaster.UCSC.dm3/man/package.html">BSgenome.Dmelanogaster.UCSC.dm3</a></span>

<span class="co"># Get sequences in a 100bp window around the peak summit</span>
<span class="no">summit_flank</span> <span class="kw">&lt;-</span> <span class="no">example_chip_summits</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_center</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">100</span>)</pre></body></html></div>
</div>
</div>
<div id="determinants-of-ectopic-and-orphan-binding" class="section level2">
<h2 class="hasAnchor">
<a href="#determinants-of-ectopic-and-orphan-binding" class="anchor"></a>Determinants of ectopic and orphan binding</h2>
<div id="pre-filtering-database-for-expressed-transcription-factors" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-filtering-database-for-expressed-transcription-factors" class="anchor"></a>Pre-filtering database for expressed transcription factors</h3>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">meme_db_path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/flyFactorSurvey_cleaned.meme"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"memes"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">universalmotif</span>)
<span class="co">#&gt; Registered S3 method overwritten by 'treeio':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   root.phylo ape</span>
<span class="no">meme_db</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/read_meme.html">read_meme</a></span>(<span class="no">meme_db_path</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/as_universalmotif_dataframe.html">as_universalmotif_dataframe</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"example_rnaseq"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"memes"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)

<span class="no">example_rnaseq</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">symbol</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">fpkm</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">fpkm</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">fpkm</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="kw">binwidth</span> <span class="kw">=</span> <span class="fl">1</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">25</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">title</span> <span class="kw">=</span> <span class="st">"Largest FPKM value in RNAseq dataset for each gene"</span>,
         <span class="kw">y</span> <span class="kw">=</span> <span class="st">"Number of genes"</span>,
         <span class="kw">x</span> <span class="kw">=</span> <span class="st">"Largest FPKM Value"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-8-1.png" width="480"></p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">expressed_genes</span> <span class="kw">&lt;-</span> <span class="no">example_rnaseq</span> <span class="kw">%&gt;%</span>
  <span class="co"># For each gene, keep only those with max(FPKM) greater or equal to 5.</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">symbol</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">fpkm</span>) <span class="kw">&gt;=</span> <span class="fl">5</span>)</pre></body></html></div>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">meme_db_expressed</span> <span class="kw">&lt;-</span> <span class="no">meme_db</span> <span class="kw">%&gt;%</span>
  <span class="co"># the altname slot of meme_db contains the gene symbol</span>
  <span class="co"># (this is database-specific)</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">altname</span> <span class="kw">%in%</span> <span class="no">expressed_genes</span>$<span class="no">symbol</span>)</pre></body></html></div>
<p>Number of motifs pre-filtering: 619</p>
<p>Number of motifs post-filter: 354</p>
<p>We’ve decreased our multiple-testing penalty by ~55%.</p>
</div>
<div id="examination-of-binding-categories-with-ame" class="section level3">
<h3 class="hasAnchor">
<a href="#examination-of-binding-categories-with-ame" class="anchor"></a>Examination of binding categories with AME</h3>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">by_binding</span> <span class="kw">&lt;-</span> <span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="fu">mcols</span>(<span class="no">.</span>)$<span class="no">peak_binding_description</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span>(<span class="no">dm.genome</span>)</pre></body></html></div>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">meme_db</span> <span class="kw">=</span> <span class="fu"><a href="../reference/as_universalmotif.html">as_universalmotif</a></span>(<span class="no">meme_db_expressed</span>))</pre></body></html></div>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">ame_by_binding</span> <span class="kw">&lt;-</span> <span class="no">by_binding</span> <span class="kw">%&gt;%</span>
  <span class="no">runAme</span></pre></body></html></div>
</div>
<div id="visualizing-ame-results" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-ame-results" class="anchor"></a>Visualizing AME results</h3>
<p>The <code><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap()</a></code> function provides a quick way to visualize AME results. It is built on top of <code>ggplot2</code>, so all <code>ggplot2</code> functions can be used to further modify the plot.</p>
<p>By default, it uses the -log10(adjusted p-value) as the heat values. See the documentation for additional notes on customization.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">ame_by_binding</span>$<span class="no">ectopic</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">rank</span> <span class="kw">%in%</span> <span class="fl">1</span>:<span class="fl">10</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group_name</span> <span class="kw">=</span> <span class="st">"Ectopic Sites"</span>)  +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Top 10 AME Hits in Ectopic Sites"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>To plot results from multiple runs together, they must first be joined into 1 data frame. The <code>ame_by_binding</code> object is a list whose names correspond to the E93 binding category. The list can be combined into a data.frame using <code><a href="https://dplyr.tidyverse.org/reference/bind.html">dplyr::bind_rows</a></code>. Setting <code>.id = "binding_type</code> creates a new column <code>binding_type</code> that contains the names from the <code>ame_by_binding</code> list. In this way, the <code>ame_res</code> data.frame contains all AME results for each run, which can be distinguished by the <code>binding_type</code> column.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">ame_res</span> <span class="kw">&lt;-</span> <span class="no">ame_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(<span class="kw">.id</span> <span class="kw">=</span> <span class="st">"binding_type"</span>)</pre></body></html></div>
<p>It is possible to aggregate results from multiple runs into a heatmap by setting the <code>group</code> parameter in <code><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap()</a></code>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">ame_res</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-16-1.png" width="1440"></p>
</div>
<div id="reducing-redundant-motif-hits" class="section level3">
<h3 class="hasAnchor">
<a href="#reducing-redundant-motif-hits" class="anchor"></a>Reducing redundant motif hits</h3>
<p>There is another key consideration for the above visualization and that is that in the FlyFactorSurvey database we used, different TFs can have multiple motif entries in the database which are all detected separately by AME. Here, when returning the top 5 hits from each group, you can see, for example, that a motif matching “Aef1” is reported 2 times within the top 5 hits of the ectopic sites (A similar result can also be seen for motifs matching “klu”). In this situation, it makes sense to summarize the data at the TF level, instead of the motif level. <strong>Note:</strong> There may be exceptions to this if for example, a TF has multiple DNA binding sequences it can recognize, in which case having multiple hits may reflect a biological property of your sequences. You will have to handle this on a case-by-case basis for interesting hits and different motif databases. Here, we can see that at least for Aef1 and klu, the consensus sequences are very similar.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">ame_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">binding_type</span> <span class="kw">==</span> <span class="st">"ectopic"</span>, <span class="no">rank</span> <span class="kw">%in%</span> <span class="fl">1</span>:<span class="fl">5</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">5</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">binding_type</span>, <span class="no">rank</span>, <span class="no">motif_id</span>, <span class="no">motif_alt_id</span>, <span class="no">consensus</span>)
<span class="co">#&gt; # A tibble: 5 x 5</span>
<span class="co">#&gt;   binding_type  rank motif_id             motif_alt_id consensus      </span>
<span class="co">#&gt;   &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;                &lt;chr&gt;        &lt;chr&gt;          </span>
<span class="co">#&gt; 1 ectopic          1 Aef1_FlyReg          Aef1         CAACWACW       </span>
<span class="co">#&gt; 2 ectopic          2 Aef1_SANGER_5        Aef1         RCAACAACAAC    </span>
<span class="co">#&gt; 3 ectopic          3 klu_SOLEXA_5         klu          KGYGKGGGTGKKDBN</span>
<span class="co">#&gt; 4 ectopic          4 l(3)neo38_SOLEXA_2.5 l(3)neo38    DGKGGGKGGGGGDGD</span>
<span class="co">#&gt; 5 ectopic          5 Mes2_SANGER_5        Mes2         TTTARGCCAAVW</span></pre></body></html></div>
<p>How to solve this problem will vary with different motif databases (For details on how to pre-process a motif database, see <code><a href="../articles/tidy_motifs.html">vignette("tidy_motifs")</a></code>). In the tidy version of the FlyFactorSurvey database, the <code>altname</code>s of the motifs are set to the transcription factor gene symbol. This information is included in the AME results as the <code>motif_alt_id</code> column.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">ame_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">binding_type</span>, <span class="no">motif_alt_id</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">adj.pvalue</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">adj.pvalue</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">motif_alt_id</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">y</span> <span class="kw">=</span> <span class="st">"Binding Category"</span>,
         <span class="kw">x</span> <span class="kw">=</span> <span class="st">"Transcription Factor Motif"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
</div>
<div id="ame-heatmap-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#ame-heatmap-visualization" class="anchor"></a>AME Heatmap Visualization</h3>
<p>If you have read <code><a href="../articles/core_ame.html">vignette("core_ame")</a></code>, the “normalized rank” heatmap was introduced as an alternative visualization method useful when AME produces very large numbers of hits, or when p-values are on very different scales between groups. The <code><a href="../reference/ame_compare_heatmap_methods.html">ame_compare_heatmap_methods()</a></code> function can be used to visualize why.</p>
<p>Below is a comparison of the distribution of values when using <code>-log10(adj.pvalue)</code> (A) vs normalized ranks (B). Because there are relatively few hits in the results (~30), and the number of hits between groups varies more than the <code>-log10(p-value)</code> distributions, the “normalize” method will produce a misleading heatmap relative to the <code>-log10(p-value)</code> map.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">ame_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">binding_type</span>, <span class="no">motif_alt_id</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">adj.pvalue</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">adj.pvalue</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_compare_heatmap_methods.html">ame_compare_heatmap_methods</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-19-1.png" width="720"></p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">best_ame_hits</span> <span class="kw">&lt;-</span> <span class="no">ame_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">binding_type</span>, <span class="no">motif_alt_id</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">adj.pvalue</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">adj.pvalue</span>))

<span class="no">pval_heatmap</span> <span class="kw">&lt;-</span> <span class="no">best_ame_hits</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">motif_alt_id</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="kw">NULL</span>,
         <span class="kw">title</span> <span class="kw">=</span> <span class="st">"-log10(adj.pvalue)"</span>)


<span class="no">norm_heatmap</span> <span class="kw">&lt;-</span> <span class="no">best_ame_hits</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">motif_alt_id</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="st">"normalize"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="kw">NULL</span>,
         <span class="kw">title</span> <span class="kw">=</span> <span class="st">"normalize"</span>)

<span class="no">pval_scaled_heatmap</span> <span class="kw">&lt;-</span> <span class="no">best_ame_hits</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">motif_alt_id</span>, <span class="kw">scale_max</span> <span class="kw">=</span> <span class="fl">10</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="kw">NULL</span>,
         <span class="kw">title</span> <span class="kw">=</span> <span class="st">"-log10(adj.pvalue) (scale capped at 10)"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="no">pval_heatmap</span>,
                   <span class="no">norm_heatmap</span>,
                   <span class="no">pval_scaled_heatmap</span>,
                   <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1</span>,
                   <span class="kw">labels</span> <span class="kw">=</span> <span class="st">"AUTO"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-21-1.png" width="1152"></p>
</div>
</div>
<div id="denovo-motif-similarity-by-binding" class="section level2">
<h2 class="hasAnchor">
<a href="#denovo-motif-similarity-by-binding" class="anchor"></a>Denovo motif similarity by binding</h2>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">dreme_by_binding</span> <span class="kw">&lt;-</span> <span class="no">by_binding</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/runDreme.html">runDreme</a></span>(<span class="st">"shuffle"</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(<span class="kw">.id</span> <span class="kw">=</span> <span class="st">"binding_type"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co"># The above code chunk takes a long time to run.</span>
<span class="co"># memes is packaged with the results of this run in the "example_dreme_by_binding" dataset</span>
<span class="co"># which can be loaded as follows:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"example_dreme_by_binding"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"memes"</span>)
<span class="no">dreme_by_binding</span> <span class="kw">&lt;-</span> <span class="no">example_dreme_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(<span class="kw">.id</span> <span class="kw">=</span> <span class="st">"binding_type"</span>)</pre></body></html></div>
<p>Rename the motifs to indicate the binding category they were discovered in.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">dreme_by_binding_renamed</span> <span class="kw">&lt;-</span> <span class="no">dreme_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">binding_type</span>, <span class="no">seq</span>, <span class="kw">sep</span> <span class="kw">=</span> <span class="st">"_"</span>)) <span class="kw">%&gt;%</span>
  <span class="co"># update_motifs updates the information in the special `motif` column</span>
  <span class="fu"><a href="../reference/update_motifs.html">update_motifs</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">cols</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"dodgerblue4"</span>))(<span class="fl">255</span>)

<span class="no">anno.df</span> <span class="kw">&lt;-</span> <span class="no">dreme_by_binding_renamed</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">name</span>, <span class="no">binding_type</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span>(<span class="st">"name"</span>)

<span class="no">dreme_by_binding_renamed</span> <span class="kw">%&gt;%</span>
  {<span class="kw pkg">universalmotif</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/compare_motifs.html">compare_motifs</a></span>(<span class="no">.</span>$<span class="no">motif</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PCC"</span>)} <span class="kw">%&gt;%</span>
  <span class="kw pkg">pheatmap</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">cols</span>,
                     <span class="kw">breaks</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">0</span>, <span class="fl">1</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="fl">1</span>/<span class="fl">255</span>),
                     <span class="kw">annotation_col</span> <span class="kw">=</span> <span class="no">anno.df</span>,
                     <span class="kw">show_colnames</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-24-1.png" width="960"></p>
<div id="test-denovo-motif-enrichment-using-ame" class="section level3">
<h3 class="hasAnchor">
<a href="#test-denovo-motif-enrichment-using-ame" class="anchor"></a>Test Denovo motif enrichment using AME</h3>
<p>The above analysis suggests the motif content of the different binding categories are highly similar. To extend these analyses, we can use AME to test for motif enrichment of the denovo discovered motifs within each binding category.</p>
<p><code><a href="../reference/runAme.html">runAme()</a></code> allows using a <code><a href="../reference/runDreme.html">runDreme()</a></code> results object as the <code>database</code> input by passing it within a <code><a href="https://rdrr.io/r/base/list.html">list()</a></code>. Naming the <code><a href="https://rdrr.io/r/base/list.html">list()</a></code> entry produces an informative <code>motif_db</code> name in the results data.frame.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">ame_denovo_by_binding</span> <span class="kw">&lt;-</span> <span class="no">by_binding</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/runAme.html">runAme</a></span>(<span class="kw">database</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"denovo_binding_motifs"</span> <span class="kw">=</span> <span class="no">dreme_by_binding_renamed</span>)) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(<span class="kw">.id</span> <span class="kw">=</span> <span class="st">"binding_type"</span>)</pre></body></html></div>
<p>Plotting the heatmap of results reveals that indeed a majority of the Denovo motifs discovered within a single category are detected in all 3 categories, supporting the conclusion that orphan, ectopic, and entopic sites are highly similar in sequence content.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">ame_denovo_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-25-1.png" width="960"></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">ame_denovo_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_compare_heatmap_methods.html">ame_compare_heatmap_methods</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-26-1.png" width="720"></p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">ame_denovo_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ame_plot_heatmap.html">ame_plot_heatmap</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">binding_type</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="st">"normalize"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-27-1.png" width="960"></p>
<p>However, there are 4 interesting motifs which distinguish orphan and ectopic sites from entopic sites. To help identify which TFs these motifs might belong to, we can use TomTom to match them to known motifs.</p>
<p>First, we isolate the motifs by ID which are unique to 1 group.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">ame_denovo_binding_unique</span> <span class="kw">&lt;-</span> <span class="no">ame_denovo_by_binding</span> <span class="kw">%&gt;%</span>
  <span class="co"># Counts the number of groups a motif is found within</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">motif_id</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">add_count</a></span>() <span class="kw">%&gt;%</span>
  <span class="co"># Selects only rows from motifs found in 1 group</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">n</span> <span class="kw">==</span> <span class="fl">1</span>)</pre></body></html></div>
<p>Next, we use the motif id’s from the AME results to filter the dreme results, and run TomTom on that subset.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">dreme_by_binding_unique</span> <span class="kw">&lt;-</span> <span class="no">dreme_by_binding_renamed</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">name</span> <span class="kw">%in%</span> <span class="no">ame_denovo_binding_unique</span>$<span class="no">motif_id</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/runTomTom.html">runTomTom</a></span>(<span class="kw">dist</span> <span class="kw">=</span> <span class="st">"ed"</span>)</pre></body></html></div>
<p>Finally, we visualize the TomTom results to identify candidate TFs driving the presence of ectopic and orphan sites.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">dreme_by_binding_unique</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/view_tomtom_hits.html">view_tomtom_hits</a></span>(<span class="fl">3</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="kw">plotlist</span> <span class="kw">=</span> <span class="no">.</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="st">"AUTO"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-30-1.png" width="1152"></p>
</div>
</div>
<div id="motifs-in-opening-vs-closing-sites" class="section level2">
<h2 class="hasAnchor">
<a href="#motifs-in-opening-vs-closing-sites" class="anchor"></a>Motifs in opening vs closing sites</h2>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="co"># split by response to E93 binding</span>
<span class="no">by_sens</span> <span class="kw">&lt;-</span> <span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="fu">mcols</span>(<span class="no">.</span>)$<span class="no">e93_sensitive_behavior</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span>(<span class="no">dm.genome</span>)</pre></body></html></div>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">dreme_by_sens_vs_static</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/runDreme.html">runDreme</a></span>(<span class="no">by_sens</span>, <span class="st">"Static"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">dreme_results</span> <span class="kw">&lt;-</span> <span class="no">dreme_by_sens_vs_static</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span>(<span class="kw">.id</span> <span class="kw">=</span> <span class="st">"e93_response"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">all_res</span> <span class="kw">&lt;-</span> <span class="no">dreme_results</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/runTomTom.html">runTomTom</a></span>(<span class="kw">dist</span> <span class="kw">=</span> <span class="st">"ed"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="no">all_res</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/view_tomtom_hits.html">view_tomtom_hits</a></span>(<span class="fl">3</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="kw">plotlist</span> <span class="kw">=</span> <span class="no">.</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="st">"AUTO"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-34-1.png" width="768"></p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="co"># use force_best_match to use update the best match info</span>
<span class="no">all_res</span> <span class="kw">%&lt;&gt;%</span>
  <span class="fu"><a href="../reference/force_best_match.html">force_best_match</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"m01_CSAAAAM"</span> <span class="kw">=</span> <span class="st">"Eip93F_SANGER_10"</span>,
                     <span class="st">"m03_AKGG"</span> <span class="kw">=</span> <span class="st">"pho_SANGER_10"</span>))</pre></body></html></div>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">all_res</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/view_tomtom_hits.html">view_tomtom_hits</a></span>(<span class="fl">1</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="kw">plotlist</span> <span class="kw">=</span> <span class="no">.</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="st">"AUTO"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-36-1.png" width="768"></p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">all_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">label</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">e93_response</span>, <span class="st">" in response to E93"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">.</span>$<span class="no">label</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">purrr</span><span class="kw ns">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span>(~{
    <span class="no">top_hits</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/view_tomtom_hits.html">view_tomtom_hits</a></span>(<span class="no">.x</span>, <span class="fl">1</span>) <span class="kw">%&gt;%</span>
      <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="kw">plotlist</span> <span class="kw">=</span> <span class="no">.</span>, <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="st">"AUTO"</span>)

    <span class="no">title</span> <span class="kw">&lt;-</span> <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/ggdraw.html">ggdraw</a></span>() +
        <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/draw_text.html">draw_text</a></span>(<span class="no">.y</span>)

    <span class="kw pkg">cowplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(<span class="kw">plotlist</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">title</span>, <span class="no">top_hits</span>),
                       <span class="kw">ncol</span> <span class="kw">=</span> <span class="fl">1</span>,
                       <span class="kw">rel_heights</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">1</span>)
    )
  })
<span class="co">#&gt; $`Decreasing in response to E93`</span></pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-37-1.png" width="768"></p>
<pre><code>#&gt; 
#&gt; $`Increasing in response to E93`</code></pre>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-37-2.png" width="768"></p>
<div id="fimo-by-response" class="section level3">
<h3 class="hasAnchor">
<a href="#fimo-by-response" class="anchor"></a>FIMO (by response)</h3>
<p>Because we discover a <em>de-novo</em> motif from DREME that matches to E93, it can be useful to scan peaks for motif occurrances to further examine properties of different peak categories.</p>
<p>Although we could also import the motifs from our local .meme database file, it is also possible to use motifs pulled from a <a href="https://bioconductor.org/packages/release/bioc/html/MotifDb.html"><code>MotifDb</code></a> query as follows.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">e93_flyfactor</span> <span class="kw">&lt;-</span> <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/MotifDb/man/MotifDb.html">MotifDb</a></span> <span class="kw">%&gt;%</span>
  <span class="co"># Query the database for the E93 motif using it's gene name</span>
  <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="st">"Eip93F"</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Convert from motifdb format to universalmotif format</span>
  <span class="kw pkg">universalmotif</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html">convert_motifs</a></span>() <span class="kw">%&gt;%</span>
  <span class="co"># The result is a list, to simplify the object, return it as a single universalmotif</span>
  <span class="no">.</span><span class="kw">[[</span><span class="fl">1</span>]]
<span class="co">#&gt; See system.file("LICENSE", package="MotifDb") for use restrictions.</span>

<span class="co"># Rename the motif from it's flybase gene number to a more user-friendly name</span>
<span class="no">e93_flyfactor</span>[<span class="st">"name"</span>] <span class="kw">&lt;-</span> <span class="st">"E93_FlyFactor"</span>

<span class="no">e93_flyfactor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        Motif name:   E93_FlyFactor</span>
<span class="co">#&gt;    Alternate name:   Eip93F_SANGER_10_FBgn0013948</span>
<span class="co">#&gt;          Organism:   Dmelanogaster</span>
<span class="co">#&gt;          Alphabet:   DNA</span>
<span class="co">#&gt;              Type:   PPM</span>
<span class="co">#&gt;           Strands:   +-</span>
<span class="co">#&gt;          Total IC:   12.12</span>
<span class="co">#&gt;         Consensus:   ACNSCCRAAAWT</span>
<span class="co">#&gt;      Target sites:   18</span>
<span class="co">#&gt;        Extra info:   [dataSource] FlyFactorSurvey </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;      A    C    N    S C C    R A A    A    W    T</span>
<span class="co">#&gt; A 0.58 0.05 0.37 0.05 0 0 0.42 1 1 0.84 0.58 0.06</span>
<span class="co">#&gt; C 0.21 0.68 0.16 0.53 1 1 0.00 0 0 0.00 0.05 0.00</span>
<span class="co">#&gt; G 0.21 0.11 0.11 0.32 0 0 0.58 0 0 0.00 0.00 0.06</span>
<span class="co">#&gt; T 0.00 0.16 0.37 0.11 0 0 0.00 0 0 0.16 0.37 0.89</span></pre></body></html></div>
<p>Just like <code><a href="../reference/runDreme.html">runDreme()</a></code> and <code><a href="../reference/runAme.html">runAme()</a></code>, <code><a href="../reference/runFimo.html">runFimo()</a></code> takes a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">Biostrings::XStringSet</a></code> as input. Using <code><a href="../reference/get_sequence.html">get_sequence()</a></code> to generate these for DNA sequences is preferred because it creates sequence inputs that can be parsed into genomic coordinates by Fimo.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">fimo_res</span> <span class="kw">&lt;-</span> <span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span>(<span class="no">dm.genome</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/runFimo.html">runFimo</a></span>(<span class="kw">motifs</span> <span class="kw">=</span> <span class="no">e93_flyfactor</span>, <span class="kw">thresh</span> <span class="kw">=</span> <span class="fl">1e-3</span>)</pre></body></html></div>
<p>The results of <code><a href="../reference/runFimo.html">runFimo()</a></code> are returned as a <code>GRanges</code> object containing the positions of each motif discovered in the input sequences. The best way to integrate these data with our input peaks is to use the <code>plyranges</code> suite of tools for performing overlaps and joins between <code>GRanges</code> objects.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">fimo_res</span>
<span class="co">#&gt; GRanges object with 1542 ranges and 6 metadata columns:</span>
<span class="co">#&gt;          seqnames            ranges strand |      motif_id</span>
<span class="co">#&gt;             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;</span>
<span class="co">#&gt;      [1]    chr3L       66285-66296      + | E93_FlyFactor</span>
<span class="co">#&gt;      [2]    chr3L       66310-66321      - | E93_FlyFactor</span>
<span class="co">#&gt;      [3]    chr3L       70891-70902      - | E93_FlyFactor</span>
<span class="co">#&gt;      [4]    chr3L     135066-135077      + | E93_FlyFactor</span>
<span class="co">#&gt;      [5]    chr3L     138527-138538      - | E93_FlyFactor</span>
<span class="co">#&gt;      ...      ...               ...    ... .           ...</span>
<span class="co">#&gt;   [1538]    chr3R 27548873-27548884      - | E93_FlyFactor</span>
<span class="co">#&gt;   [1539]    chr3R 27754398-27754409      - | E93_FlyFactor</span>
<span class="co">#&gt;   [1540]    chr3R 27754407-27754418      + | E93_FlyFactor</span>
<span class="co">#&gt;   [1541]    chr3R 27780275-27780286      + | E93_FlyFactor</span>
<span class="co">#&gt;   [1542]    chr3R 27853336-27853347      - | E93_FlyFactor</span>
<span class="co">#&gt;                          motif_alt_id     score    pvalue    qvalue</span>
<span class="co">#&gt;                           &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;logical&gt;</span>
<span class="co">#&gt;      [1] Eip93F_SANGER_10_FBgn0013948      5.22  0.000436      &lt;NA&gt;</span>
<span class="co">#&gt;      [2] Eip93F_SANGER_10_FBgn0013948     14.25  1.01e-05      &lt;NA&gt;</span>
<span class="co">#&gt;      [3] Eip93F_SANGER_10_FBgn0013948      5.64  0.000379      &lt;NA&gt;</span>
<span class="co">#&gt;      [4] Eip93F_SANGER_10_FBgn0013948       4.4  0.000571      &lt;NA&gt;</span>
<span class="co">#&gt;      [5] Eip93F_SANGER_10_FBgn0013948      10.1  8.96e-05      &lt;NA&gt;</span>
<span class="co">#&gt;      ...                          ...       ...       ...       ...</span>
<span class="co">#&gt;   [1538] Eip93F_SANGER_10_FBgn0013948      5.57  0.000391      &lt;NA&gt;</span>
<span class="co">#&gt;   [1539] Eip93F_SANGER_10_FBgn0013948      7.37  0.000222      &lt;NA&gt;</span>
<span class="co">#&gt;   [1540] Eip93F_SANGER_10_FBgn0013948      7.84  0.000192      &lt;NA&gt;</span>
<span class="co">#&gt;   [1541] Eip93F_SANGER_10_FBgn0013948      8.75  0.000145      &lt;NA&gt;</span>
<span class="co">#&gt;   [1542] Eip93F_SANGER_10_FBgn0013948      4.59  0.000536      &lt;NA&gt;</span>
<span class="co">#&gt;          matched_sequence</span>
<span class="co">#&gt;                 &lt;logical&gt;</span>
<span class="co">#&gt;      [1]             &lt;NA&gt;</span>
<span class="co">#&gt;      [2]             &lt;NA&gt;</span>
<span class="co">#&gt;      [3]             &lt;NA&gt;</span>
<span class="co">#&gt;      [4]             &lt;NA&gt;</span>
<span class="co">#&gt;      [5]             &lt;NA&gt;</span>
<span class="co">#&gt;      ...              ...</span>
<span class="co">#&gt;   [1538]             &lt;NA&gt;</span>
<span class="co">#&gt;   [1539]             &lt;NA&gt;</span>
<span class="co">#&gt;   [1540]             &lt;NA&gt;</span>
<span class="co">#&gt;   [1541]             &lt;NA&gt;</span>
<span class="co">#&gt;   [1542]             &lt;NA&gt;</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></pre></body></html></div>
</div>
<div id="counting-the-number-of-motifs-per-peak" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-the-number-of-motifs-per-peak" class="anchor"></a>Counting the number of motifs per peak</h3>
<p><code>plyranges</code> extends <code>dplyr</code>-like syntax to range objects. Here we add a count for each motif per peak in the <code>n_motifs</code> column. We also add a column <code>has_motif</code> which will be a binary indicator of whether a peak contains any motifs.</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">summit_flank</span> <span class="kw">%&lt;&gt;%</span>
  <span class="co"># Note, if running FIMO with multiple motifs, this solution will not work</span>
  <span class="co"># as it will count all motifs within the fimo-results without splitting by motif_id</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">n_motifs</span> <span class="kw">=</span> <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html">count_overlaps</a></span>(<span class="no">.</span>, <span class="no">fimo_res</span>),
                    <span class="kw">has_motif</span> <span class="kw">=</span> <span class="no">n_motifs</span> <span class="kw">&gt;</span> <span class="fl">0</span>)</pre></body></html></div>
<p>First, to we want to determine whether E93 sensitive sites are more likely to have E93 motifs in certain response types. Here we can see that sensitive decreasing sites are more likely to have E93 motifs than sensitive increasing or insensitive static sites.</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">has_match</span> <span class="kw">=</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span>(<span class="no">has_motif</span>, <span class="st">"Match"</span>, <span class="st">"No Match"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">e93_sensitive_behavior</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="kw pkg">forcats</span><span class="kw ns">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_rev.html">fct_rev</a></span>(<span class="no">has_match</span>)), <span class="kw">position</span> <span class="kw">=</span> <span class="st">"fill"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Match"</span> <span class="kw">=</span> <span class="st">"firebrick"</span>,
                                 <span class="st">"No Match"</span> <span class="kw">=</span> <span class="st">"Black"</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="st">"E93 Motif Match"</span>,
         <span class="kw">y</span> <span class="kw">=</span> <span class="st">"Fraction of Sites"</span>,
         <span class="kw">x</span> <span class="kw">=</span> <span class="st">"Response to E93 binding"</span>)</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-42-1.png" width="480"></p>
<p>To investigate whether E93-sensitive sites that open/close/don’t change have different numbers of E93 motifs, we plot the fraction of sites with each number of motifs. Here we can see that in addition to being more likely to contain an E93 motif, sensitive decreasing sites are more likely to to contain 2 or more matches, where 10% contain at least 2 motifs.</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="co"># currently, group operations are faster as data.frames, so we convert to data.frame</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">e93_sensitive_behavior</span>, <span class="no">n_motifs</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">e93_sensitive_behavior</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">frac</span> <span class="kw">=</span> <span class="no">n</span>/<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">n</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">n_motifs</span>, <span class="no">frac</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">e93_sensitive_behavior</span>), <span class="kw">size</span> <span class="kw">=</span> <span class="fl">1</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">y</span> <span class="kw">=</span> <span class="st">"Fraction of Sites"</span>,
         <span class="kw">x</span> <span class="kw">=</span> <span class="st">"Number of E93 Motifs"</span>,
         <span class="kw">color</span> <span class="kw">=</span> <span class="st">"Response to E93 Binding"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-43-1.png" width="672"></p>
<p>Finally, we want to assess whether the quality of E93 motifs is different between sensitivity categories. To examine this, we need to determine which motifs are found in which peaks. We use the <code><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">plyranges::join_overlap_intersect</a></code> to return motif entries appended with peak-level metadata, like the peak id each motif is found within.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="co"># return position of each motif match w/ peak metadata</span>
<span class="no">intersect</span> <span class="kw">&lt;-</span> <span class="no">fimo_res</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_intersect</a></span>(<span class="no">summit_flank</span>)</pre></body></html></div>
<p>We use the FIMO <code>score</code> as a proxy for quality, where higher scores are better matches to the motif. Here we examine only the best match (highest score) motif within each peak.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="no">best_motifs</span> <span class="kw">&lt;-</span> <span class="no">intersect</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">group_by</a></span>(<span class="no">id</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Keep only motifs w/ the highest score within a peak</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span>(<span class="no">score</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">score</span>)) <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">ungroup</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="no">best_motifs</span> <span class="kw">%&gt;%</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">e93_sensitive_behavior</span>, <span class="no">score</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="no">e93_sensitive_behavior</span>), <span class="kw">notch</span> <span class="kw">=</span> <span class="no">T</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.5</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="kw">fill</span> <span class="kw">=</span> <span class="st">"none"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"Response to E93"</span>,
         <span class="kw">y</span> <span class="kw">=</span> <span class="st">"FIMO Score"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-46-1.png" width="288"></p>
<p>The <a href="https://en.wikipedia.org/wiki/Anderson%E2%80%93Darling_test">Anderson-Darling test</a> can be used to test differences in the distribution of FIMO scores across multiple groups. For a pairwise comparison, a <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">Kolmogorov-Smirnov test</a> is commonly used instead.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="no">best_motifs</span> <span class="kw">%&gt;%</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">behavior</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">e93_sensitive_behavior</span>)) <span class="kw">%&gt;%</span>
  <span class="kw pkg">PMCMRplus</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/PMCMRplus/man/adAllPairsTest.html">adAllPairsTest</a></span>(<span class="no">score</span> ~ <span class="no">behavior</span>, <span class="no">.</span>, <span class="kw">p.adjust.method</span> <span class="kw">=</span> <span class="st">"fdr"</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;  Pairwise comparisons using Anderson-Darling All-Pairs Test</span>
<span class="co">#&gt; data: score by behavior</span>
<span class="co">#&gt;            Decreasing Increasing</span>
<span class="co">#&gt; Increasing 1.5e-05    -         </span>
<span class="co">#&gt; Static     0.00011    0.02132</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; P value adjustment method: fdr</span>
<span class="co">#&gt; alternative hypothesis: two.sided</span></pre></body></html></div>
<p>We can visualize the differences in E93 motif quality between groups by reconstructing PCM’s from the matched motif sequences. Here, it is clear that although each category contains an E93 motif match, the sequence content of those motifs differs between categories, providing a nice visual representation of the statistical analysis above.</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="no">best_motifs</span> <span class="kw">%&gt;%</span>
  <span class="co"># Get sequences of each matched motif</span>
  <span class="fu"><a href="../reference/add_sequence.html">add_sequence</a></span>(<span class="no">dm.genome</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Split by E93 response category</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="fu">mcols</span>(<span class="no">.</span>)$<span class="no">e93_sensitive_behavior</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Convert from GRangesList to list() to use `purrr::imap`</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>() <span class="kw">%&gt;%</span>
  <span class="co"># imap passes the list entry as .x and the name of that object to .y</span>
  <span class="kw pkg">purrr</span><span class="kw ns">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span>(~{
    <span class="co"># run create_motif() on each set of sequences</span>
    <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/create_motif.html">create_motif</a></span>(<span class="no">.x</span>$<span class="no">sequence</span>,
                 <span class="co"># Append the response category to the motif name</span>
                 <span class="kw">name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"E93_"</span>, <span class="no">.y</span>))
    }) <span class="kw">%&gt;%</span>
  <span class="co"># Prepend E93 Fly Factor motif to beginning of motif list</span>
  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"E93_FlyFactor"</span> <span class="kw">=</span> <span class="no">e93_flyfactor</span>), <span class="no">.</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Plot each motif</span>
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html">view_motifs</a></span>()</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-48-1.png" width="480"></p>
</div>
<div id="centrality-of-e93-motif" class="section level3">
<h3 class="hasAnchor">
<a href="#centrality-of-e93-motif" class="anchor"></a>Centrality of E93 motif</h3>
<p>Next we want to visualize whether the E93 motif has altered positioning across the response categories. To do this we add the metadata for each nearest motif to our peak summits. Setting <code>distance = TRUE</code> in <code><a href="https://rdrr.io/pkg/plyranges/man/ranges-nearest.html">plyranges::join_nearest</a></code> adds a <code>distance</code> column indicating the distance to the nearest joined region.</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="no">summit_nearest_e93_motif</span> <span class="kw">&lt;-</span> <span class="no">summit_flank</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html">anchor_center</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">mutate</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">1</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">plyranges</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-nearest.html">join_nearest</a></span>(<span class="no">fimo_res</span>, <span class="kw">distance</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="no">summit_nearest_e93_motif</span> <span class="kw">%&gt;%</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">has_motif</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">distance</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ecdf.html">stat_ecdf</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">e93_sensitive_behavior</span>), <span class="kw">size</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">pad</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"Distance to nearest E93 Motif"</span>,
         <span class="kw">y</span> <span class="kw">=</span> <span class="st">"Fraction of Sites"</span>,
         <span class="kw">color</span> <span class="kw">=</span> <span class="st">"E93 Response"</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_linedraw</a></span>()</pre></body></html></div>
<p><img src="integrative_analysis_files/figure-html/unnamed-chunk-50-1.png" width="700"></p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="no">summit_nearest_e93_motif</span> <span class="kw">%&gt;%</span>
  <span class="no">data.frame</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">has_motif</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">behavior</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">e93_sensitive_behavior</span>)) <span class="kw">%&gt;%</span>
  <span class="co"># Anderson-darling test for multiple distributions</span>
  <span class="kw pkg">PMCMRplus</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/PMCMRplus/man/adAllPairsTest.html">adAllPairsTest</a></span>(<span class="no">distance</span> ~ <span class="no">behavior</span>, <span class="no">.</span>, <span class="kw">p.adjust.method</span> <span class="kw">=</span> <span class="st">"fdr"</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;  Pairwise comparisons using Anderson-Darling All-Pairs Test</span>
<span class="co">#&gt; data: distance by behavior</span>
<span class="co">#&gt;            Decreasing Increasing</span>
<span class="co">#&gt; Increasing 0.058      -         </span>
<span class="co">#&gt; Static     0.021      0.945</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; P value adjustment method: fdr</span>
<span class="co">#&gt; alternative hypothesis: two.sided</span></pre></body></html></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
