<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tidying Motif Metadata • dremeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tidying Motif Metadata">
<meta property="og:description" content="dremeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dremeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Old_Readme.html">readme</a>
    </li>
    <li>
      <a href="../articles/dreme.html">Denovo Motif Discovery Using DREME</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">Integrative Analysis</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Tidying Motif Metadata</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/dremeR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tidying Motif Metadata</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/dremeR/blob/master/vignettes/tidy_motifs.Rmd"><code>vignettes/tidy_motifs.Rmd</code></a></small>
      <div class="hidden name"><code>tidy_motifs.Rmd</code></div>

    </div>

    
    
<p>dremeR contains several experimental features to simplify working with motif metadata. These can be used to manipulate motifs detected from MEME Suite tools, or to deal with motifs from other sources.</p>
<p><strong>Note:</strong> These features are experimental and will be influenced by user feedback. Please provide comments, issues, or other feedback at the <a href="https://github.com/snystrom/dremeR/issues">dremeR issues page</a>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dremeR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">universalmotif</span>)
<span class="co">#&gt; Registered S3 method overwritten by 'treeio':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   root.phylo ape</span></pre></body></html></div>
<p>The <code>MotifDb</code> package makes it easy to query thousands of motifs from public databases. Here, I will describe how to use one of these queries as input to dremeR functions, and how to manipulate the resulting motifs to prepare them for MEME suite tools.</p>
<p>I will use the motifs from the <a href="https://mccb.umassmed.edu/ffs/">FlyFactorSurvey</a> as an example. They can be accessed from <code>MotifDb</code> using the following query.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">flyFactorDb</span> <span class="kw">&lt;-</span> <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/MotifDb/man/MotifDb.html">MotifDb</a></span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">MotifDb</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html">query</a></span>(<span class="st">"FlyFactorSurvey"</span>)
<span class="co">#&gt; See system.file("LICENSE", package="MotifDb") for use restrictions.</span></pre></body></html></div>
<p>Use <code><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html">universalmotif::convert_motifs()</a></code> to convert a <code>MotifDb</code> query into motif objects. In many cases, the resulting list can be used directly as input to dremeR functions, like <code>runTomTom</code>, or <code>runAme</code>.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">flyFactorMotifs</span> <span class="kw">&lt;-</span> <span class="no">flyFactorDb</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html">convert_motifs</a></span>()</pre></body></html></div>
<p>But there are some issues with this database. For example, the following motif name is the FlyBase gene number, and the alternate name is the actual informative name of the PWM.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">flyFactorMotifs</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">1</span>)
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        Motif name:   FBgn0259750</span>
<span class="co">#&gt;    Alternate name:   ab_SANGER_10_FBgn0259750</span>
<span class="co">#&gt;          Organism:   Dmelanogaster</span>
<span class="co">#&gt;          Alphabet:   DNA</span>
<span class="co">#&gt;              Type:   PPM</span>
<span class="co">#&gt;           Strands:   +-</span>
<span class="co">#&gt;          Total IC:   15.01</span>
<span class="co">#&gt;         Consensus:   BWNRCCAGGWMCNNTNNNGNN</span>
<span class="co">#&gt;      Target sites:   20</span>
<span class="co">#&gt;        Extra info:   [dataSource] FlyFactorSurvey </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     B    W    N    R C C A G G    W    M    C    N    N    T    N    N    N</span>
<span class="co">#&gt; A 0.0 0.50 0.20 0.35 0 0 1 0 0 0.55 0.35 0.05 0.20 0.45 0.20 0.10 0.40 0.40</span>
<span class="co">#&gt; C 0.3 0.15 0.25 0.00 1 1 0 0 0 0.10 0.65 0.70 0.45 0.25 0.10 0.25 0.25 0.10</span>
<span class="co">#&gt; G 0.4 0.05 0.50 0.65 0 0 0 1 1 0.00 0.00 0.05 0.05 0.15 0.05 0.20 0.05 0.15</span>
<span class="co">#&gt; T 0.3 0.30 0.05 0.00 0 0 0 0 0 0.35 0.00 0.20 0.30 0.15 0.65 0.45 0.30 0.35</span>
<span class="co">#&gt;      G    N    N</span>
<span class="co">#&gt; A 0.25 0.50 0.30</span>
<span class="co">#&gt; C 0.10 0.25 0.25</span>
<span class="co">#&gt; G 0.55 0.15 0.45</span>
<span class="co">#&gt; T 0.10 0.10 0.00</span></pre></body></html></div>
<p>dremeR provides the <code><a href="../reference/as_universalmotif_dataframe.html">as_universalmotif_dataframe()</a></code> method for converting universalmotif lists into special data.frames which can be used to update motif entries. This is particularly useful when dealing with several motifs at once.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">&lt;-</span> <span class="no">flyFactorMotifs</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/as_universalmotif_dataframe.html">as_universalmotif_dataframe</a></span>()</pre></body></html></div>
<p>The universalmotif data.frame contains all the columns that can be used to update motif entries. Any additional columns added to the data.frame will be ignored. The <code>motif</code> column is a list column that holds the original universalmotif objects.</p>
<p><strong>Note:</strong> If the <code>motif</code> column is altered by non-dremeR methods, the following approaches are not guaranteed to work. These features are experimental and will be influenced by user feedback. Please provide comments, issues, or other feedback at the <a href="https://github.com/snystrom/dremeR/issues">dremeR issues page</a>.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># The following columns can be used to update motif metadata</span>
<span class="no">flyFactor_data</span> <span class="kw">%&gt;%</span>
  <span class="no">names</span>
<span class="co">#&gt;  [1] "name"      "altname"   "family"    "organism"  "consensus" "alphabet" </span>
<span class="co">#&gt;  [7] "strand"    "icscore"   "nsites"    "bkgsites"  "pval"      "qval"     </span>
<span class="co">#&gt; [13] "eval"      "motif"</span></pre></body></html></div>
<p>using the universalmotif data.frame, we can quickly see that the issue with FBgn numbers only applies to certain entries. And TFs which are represented by multiple motifs in the database are assigned the same name. The MEME Suite tools which use a motif database (like TomTom and AME) require that the entries have unique primary identifiers, therefore the default names will not be appropriate.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">%&gt;%</span>
  <span class="co"># The `motif` column can slow down printing, so it can be helpful to drop it.</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(-<span class="no">motif</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">5</span>)
<span class="co">#&gt;          name                  altname family      organism</span>
<span class="co">#&gt; 1 FBgn0259750 ab_SANGER_10_FBgn0259750   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 2 FBgn0259750  ab_SOLEXA_5_FBgn0259750   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 3       abd-A Abd-A_FlyReg_FBgn0000014   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 4       Abd-B Abd-B_FlyReg_FBgn0000015   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 5       abd-A    AbdA_Cell_FBgn0000014   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites bkgsites pval</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20       NA   NA</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446       NA   NA</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37       NA   NA</span>
<span class="co">#&gt; 4                   YCATAAAA      DNA     +-  7.831193      7       NA   NA</span>
<span class="co">#&gt; 5                    TTAATKA      DNA     +-  8.976569     18       NA   NA</span>
<span class="co">#&gt;   qval eval</span>
<span class="co">#&gt; 1   NA   NA</span>
<span class="co">#&gt; 2   NA   NA</span>
<span class="co">#&gt; 3   NA   NA</span>
<span class="co">#&gt; 4   NA   NA</span>
<span class="co">#&gt; 5   NA   NA</span></pre></body></html></div>
<p>However, the <code>altname</code> slots from the <code>motifDb</code> query are already unique, so we can make them the primary name.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">flyFactor_data</span>$<span class="no">altname</span> <span class="kw">%&gt;%</span> <span class="no">length</span>
<span class="co">#&gt; [1] 614</span>
<span class="no">flyFactor_data</span>$<span class="no">altname</span> <span class="kw">%&gt;%</span> <span class="no">unique</span> <span class="kw">%&gt;%</span> <span class="no">length</span>
<span class="co">#&gt; [1] 614</span></pre></body></html></div>
<p>An easy way is to use <code><a href="https://dplyr.tidyverse.org/reference/rename.html">dplyr::rename</a></code> to swap the columns.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">%&lt;&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span>(<span class="st">"altname"</span> <span class="kw">=</span> <span class="st">"name"</span>,
                <span class="st">"name"</span> <span class="kw">=</span> <span class="st">"altname"</span>)</pre></body></html></div>
<p>The <code>name</code> column now contains the full motif name.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">3</span>)
<span class="co">#&gt;       altname                     name family      organism</span>
<span class="co">#&gt; 1 FBgn0259750 ab_SANGER_10_FBgn0259750   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 2 FBgn0259750  ab_SOLEXA_5_FBgn0259750   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 3       abd-A Abd-A_FlyReg_FBgn0000014   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites bkgsites pval</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20       NA   NA</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446       NA   NA</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37       NA   NA</span>
<span class="co">#&gt;   qval eval</span>
<span class="co">#&gt; 1   NA   NA</span>
<span class="co">#&gt; 2   NA   NA</span>
<span class="co">#&gt; 3   NA   NA</span>
<span class="co">#&gt;                                                                  motif</span>
<span class="co">#&gt; 1 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span>
<span class="co">#&gt; 2 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span>
<span class="co">#&gt; 3 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span></pre></body></html></div>
<p>Next to solve the issue with the FBgn’s. As of this writing in May 2020, the FBgn entries are out of date, making them difficult to look up within R (<a href="flybase.org">FlyBase.org</a> has a nice conversion tool for this, however). For this example, we will try to grab the correct gene name from the motif name, which is stored in the first field of the name, formatted as follows: “<gene><em><sequencing-platform></sequencing-platform></em><fbgn>”.</fbgn></gene></p>
<p>We use <code><a href="https://tidyr.tidyverse.org/reference/separate.html">tidyr::separate</a></code> to split out the first entry to the <code>tifd</code> column, then only use this value if the altname contains an FBgn.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">%&lt;&gt;%</span>
  <span class="co"># Critical to set remove = FALSE to keep the `name` column</span>
  <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(<span class="no">name</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"tfid"</span>), <span class="kw">remove</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="co"># Only use the tfid if the altname contains an FBgn</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">altname</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"^FBgn"</span>, <span class="no">altname</span>), <span class="no">tfid</span>, <span class="no">altname</span>))
<span class="co">#&gt; Warning: Expected 1 pieces. Additional pieces discarded in 614 rows [1, 2, 3, 4,</span>
<span class="co">#&gt; 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].</span></pre></body></html></div>
<p>Now, the first two entries are listed as “ab” instead of “FBgn0259750”.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">flyFactor_data</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">3</span>)
<span class="co">#&gt;   altname                     name tfid family      organism</span>
<span class="co">#&gt; 1      ab ab_SANGER_10_FBgn0259750   ab   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 2      ab  ab_SOLEXA_5_FBgn0259750   ab   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt; 3   abd-A Abd-A_FlyReg_FBgn0000014  Abd   &lt;NA&gt; Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites bkgsites pval</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20       NA   NA</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446       NA   NA</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37       NA   NA</span>
<span class="co">#&gt;   qval eval</span>
<span class="co">#&gt; 1   NA   NA</span>
<span class="co">#&gt; 2   NA   NA</span>
<span class="co">#&gt; 3   NA   NA</span>
<span class="co">#&gt;                                                                  motif</span>
<span class="co">#&gt; 1 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span>
<span class="co">#&gt; 2 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span>
<span class="co">#&gt; 3 &lt;S4 class 'universalmotif' [package "universalmotif"] with 20 slots&gt;</span></pre></body></html></div>
<p>Finally, the resulting data.frame can be converted back into a universalmotif list using <code><a href="../reference/as_universalmotif.html">as_universalmotif()</a></code>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">flyFactorMotifs_final</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/as_universalmotif.html">as_universalmotif</a></span>(<span class="no">flyFactor_data</span>)</pre></body></html></div>
<p>The resulting universalmotif list object now reflects the changes we made to the <code>data.frame</code> and can now be exported as a .meme format file using <code><a href="https://rdrr.io/pkg/universalmotif/man/write_meme.html">universalmotif::write_meme</a></code> or can be used directly as input to tools like <code>runTomTom</code> or <code>runAme</code>.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">flyFactorMotifs_final</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">1</span>)
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        Motif name:   ab_SANGER_10_FBgn0259750</span>
<span class="co">#&gt;    Alternate name:   ab</span>
<span class="co">#&gt;          Organism:   Dmelanogaster</span>
<span class="co">#&gt;          Alphabet:   DNA</span>
<span class="co">#&gt;              Type:   PPM</span>
<span class="co">#&gt;           Strands:   +-</span>
<span class="co">#&gt;          Total IC:   15.01</span>
<span class="co">#&gt;         Consensus:   BWNRCCAGGWMCNNTNNNGNN</span>
<span class="co">#&gt;      Target sites:   20</span>
<span class="co">#&gt;        Extra info:   [dataSource] FlyFactorSurvey </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     B    W    N    R C C A G G    W    M    C    N    N    T    N    N    N</span>
<span class="co">#&gt; A 0.0 0.50 0.20 0.35 0 0 1 0 0 0.55 0.35 0.05 0.20 0.45 0.20 0.10 0.40 0.40</span>
<span class="co">#&gt; C 0.3 0.15 0.25 0.00 1 1 0 0 0 0.10 0.65 0.70 0.45 0.25 0.10 0.25 0.25 0.10</span>
<span class="co">#&gt; G 0.4 0.05 0.50 0.65 0 0 0 1 1 0.00 0.00 0.05 0.05 0.15 0.05 0.20 0.05 0.15</span>
<span class="co">#&gt; T 0.3 0.30 0.05 0.00 0 0 0 0 0 0.35 0.00 0.20 0.30 0.15 0.65 0.45 0.30 0.35</span>
<span class="co">#&gt;      G    N    N</span>
<span class="co">#&gt; A 0.25 0.50 0.30</span>
<span class="co">#&gt; C 0.10 0.25 0.25</span>
<span class="co">#&gt; G 0.55 0.15 0.45</span>
<span class="co">#&gt; T 0.10 0.10 0.00</span></pre></body></html></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
