<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run FIMO on input sequences — runFimo • dremeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run FIMO on input sequences — runFimo" />
<meta property="og:description" content="NOTE: runs best using `text = TRUE`, setting genomic coordinates in fasta
headers, and using `parse_genomic_coord = TRUE`." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dremeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Matching</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">Integrative Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/snystrom/dremeR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run FIMO on input sequences</h1>
    <small class="dont-index">Source: <a href='https://github.com/snystrom/dremeR/blob/master/R/fimo.R'><code>R/fimo.R</code></a></small>
    <div class="hidden name"><code>runFimo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>NOTE: runs best using `text = TRUE`, setting genomic coordinates in fasta
headers, and using `parse_genomic_coord = TRUE`.</p>
    </div>

    <pre class="usage"><span class='fu'>runFimo</span>(
  <span class='no'>sequences</span>,
  <span class='no'>motifs</span>,
  <span class='kw'>bfile</span> <span class='kw'>=</span> <span class='st'>"motif"</span>,
  <span class='kw'>outdir</span> <span class='kw'>=</span> <span class='st'>"auto"</span>,
  <span class='kw'>parse_genomic_coord</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>skip_matched_sequence</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>max_strand</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>text</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>meme_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>silent</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sequences</th>
      <td><p>path to fasta file, or stringset input.</p></td>
    </tr>
    <tr>
      <th>motifs</th>
      <td><p>path to .meme format file, or universalmotif/universalmotif list input.</p></td>
    </tr>
    <tr>
      <th>bfile</th>
      <td><p>path to background file, or special values: "motif", or
"uniform" to use 0-order frequencies contained in the motif, or a uniform
letter distribution. (default: "motif")</p></td>
    </tr>
    <tr>
      <th>outdir</th>
      <td><p>output directory location. Only used if text = FALSE. Default:
"auto" to autogenerate directory name. Note: if not using a fasta path as
input, this will be a temporary location unless explicity set.</p></td>
    </tr>
    <tr>
      <th>parse_genomic_coord</th>
      <td><p>`logical(1)` whether to parse genomic position
from fasta name. Fasta names must be UCSC format positions (ie
"chr:start-end"). If names of fasta entries are genomic coordinates and
parse_genomic_coord == TRUE, results will contain positions of motif matches,
otherwise FIMO will count positions from 1 to length of fasta file.</p></td>
    </tr>
    <tr>
      <th>skip_matched_sequence</th>
      <td><p>`logical(1)` whether or not to include the DNA
sequence of the match. Default: `FALSE`. Note: jobs will complete faster if
set to FALSE. Other utilities can be used to lookup the sequence if
`parse_genomic_coord` is `TRUE`.</p></td>
    </tr>
    <tr>
      <th>max_strand</th>
      <td><p>if match is found on both strands, only report strand with
best match.</p></td>
    </tr>
    <tr>
      <th>text</th>
      <td><p>`logical(1)` (default: `TRUE`). No output files will be created on the filesystem.
The results are unsorted and no q-values are computed. This setting allows
fast searches on very large inputs.</p></td>
    </tr>
    <tr>
      <th>meme_path</th>
      <td><p>path to `meme/bin/` (optional). Defaut: `NULL`, searches
"MEME_PATH" environment variable or "meme_path" option.</p></td>
    </tr>
    <tr>
      <th>silent</th>
      <td><p>`logical(1)` whether to suppress stdout/stderr printing to
terminal (default: TRUE). NOTE: if `text = TRUE`, setting `silent = TRUE`
will print all FIMO matches to terminal if the run is successful. Can be
useful for troubleshooting.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional commandline arguments to fimo see:
<code>http://meme-suite.org/doc/fimo.html?man_type=web</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>GenomicRanges object contining positions of each match. Note: if
  `parse_genomic_coords = FALSE`, each `seqnames` entry will be the fasta
  header, and start/end will be the position within that sequence of the
  match. It is a good idea to use `parse_genomic_coords = TRUE`.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>seq</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>universalmotif</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/universalmotif/man/create_sequences.html'>create_sequences</a></span>()
<span class='co'># sequences must have names in their fasta headers</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>seq</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span>(<span class='no'>seq</span>)
<span class='no'>motif</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>universalmotif</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/universalmotif/man/create_motif.html'>create_motif</a></span>()
<span class='fu'>runFimo</span>(<span class='no'>seq</span>, <span class='no'>motif</span>, <span class='kw'>parse_genomic_coord</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


